src/corppa/poetry_detection/annotation/recipe.py

```diff
@@ -234,12 +243,106 @@ def annotate_page_text(
+    """
+    Extract session name from task examplke. Session ids have the following form:
+        [session id] = [db id]-[session name]
+    Assumes that session names do not contain dashes.
+    """
+    session_id = example[SESSION_ID_ATTR]
+    session_name = session_id.rsplit("-", maxsplit=1)[-1]
+    if suffix is not None:
+        session_name = f"{session_name}-{suffix}"
+    return session_name
+
+
+def add_label_prefix(label: str, prefix: str) -> str:
+    return f"{prefix}: {label}"
+
+
+def remove_label_prefix(label: str) -> str:
+    return label.rsplit(": ", maxsplit=1)[-1]
+
+
+def add_session_prefix(
+    example: TaskType, session_sfx: Optional[str] = None
+) -> TaskType:
+    """
+    Add session name as prefix to text span labels
+
+    label --> "[session]: label"
+    ex. "POETRY" --> "alice: POETRY"
+
+    Calls: `get_sesssion_name` and `add_label_prefix`
+    """
+    session_name = get_session_name(example, suffix=session_sfx)
+    for span in example.get("spans", []):
+        span["label"] = add_label_prefix(span["label"], session_name)
+    return example
+
+
+def remove_session_prefix(example: TaskType) -> TaskType:
+    """
+    Remove added session prefix from text span labels
+
+    Calls: `remove_label_prefix`
+    """
+    for span in example.get("spans", []):
+        span["label"] = remove_label_prefix(span["label"])
+    return example
+
+
+def has_span_overlap(example: TaskType, strip_label_pfx: bool = True) -> bool:
+    """
+    Check if example has overlapping (text) span annotations
+
+    Calls: `remove_label_prefix` (optionally)
+    """
+    if "spans" not in example:
+        return False
+    label_coverage = {}  # label (str) --> coverage (intspan)
+    for span in example["spans"]:
+        # Get label
+        label = span["label"]
+        if strip_label_pfx:
+            label = remove_label_prefix(label)
+        # Represent span's coverage as an intspan
+        # Note: intspans are closed intervals, while span character ranges are half-open
+        span_coverage = intspan.from_range(span["start"], span["end"] - 1)
+        # Check if label's intspan is disjoint from
+        if label not in label_coverage:
+            label_coverage[label] = span_coverage
+        else:
+            if span_coverage.isdisjoint(label_coverage[label]):
+                # No overlap, accumulate into label's coverage
+                label_coverage[label] |= span_coverage
+            else:
+                # Overlap found
+                return True
+    return False
+
+
+def validate_review_answer(example: TaskType):
+    """
+    Validate review answer:
+    * Example not flagged (i.e. flagged field not set)
+    * Text spans (by label) must not overlap
+    """
+    if example.get("flagged") is True:
+        raise ValueError("Currently flagged, unflag to submit.")
+    if has_span_overlap(example):
+        raise ValueError("Overlapping spans with the same label detected!")
```

Comment:
I think I understand the logic of the span overlap check but I'm missing why it's needed - is this something you and Mary agreed on? validation/adjudication requires cleaning up existing annotations so there is no overlap and one definitive set of spans?

Reply:
This is a customizable feature of Prodigy Recipes, to allow validation of answers *before* allowing an annotator to submit the annotation (see [documentation](https://prodi.gy/docs/custom-recipes#validate_answer)).

So, the `validate_review_answer` method is being used to validate the annotations being "submitted" in the interface (i.e., selecting the accept or reject buttons). When an error is raised, a pop-up window appears telling the annotator they cannot submit this "invalid" annotation. The error messages are also shown within this pop-up window.

Reply:
I assumed it was something like that, although I didn't know about the UI integration - seems like a very powerful feature. My question was actually that I don't understand why overlapping spans  == invalid

Reply:
So, this is to prevent duplicate and/or conflicting spans to exist when adjudicating. For our task, spans of the same label should not overlap (a passage can only be an excerpt of a single poem).

Reply:
This is primarily a validation check for the adjudicator, that they've culled enough of the existing annotations before submission.